src/blob_storage.cc

```diff
@@ -6,24 +6,71 @@
 namespace rocksdb {
 namespace titandb {
 
+std::string BlobStorage::EncodeBlobCache(const BlobIndex& index) {
+  std::string cache_key(cache_prefix_);
+  PutVarint64(&cache_key, index.file_number);
+  PutVarint64(&cache_key, index.blob_handle.offset);
+  return cache_key;
+}
+
+Status BlobStorage::TryGetBlobCache(const std::string& cache_key,
+                                    BlobRecord* record, PinnableSlice* value,
+                                    bool* cache_hit) {
+  Status s;
+  Cache::Handle* cache_handle = blob_cache_->Lookup(cache_key);
+  if (cache_handle) {
+    *cache_hit = true;
+    RecordTick(statistics(stats_), TITAN_BLOB_CACHE_HIT);
+    auto blob = reinterpret_cast<OwnedSlice*>(blob_cache_->Value(cache_handle));
+    s = DecodeInto(*blob, record);
+    if (!s.ok()) return s;
+
+    value->PinSlice(record->value, UnrefCacheHandle, blob_cache_.get(),
+                    cache_handle);
+    return s;
+  }
+  *cache_hit = false;
+  RecordTick(statistics(stats_), TITAN_BLOB_CACHE_MISS);
+  return s;
+}
+
 Status BlobStorage::Get(const ReadOptions& options, const BlobIndex& index,
                         BlobRecord* record, PinnableSlice* value) {
-  auto sfile = FindFile(index.file_number).lock();
-  if (!sfile)
-    return Status::Corruption("Missing blob file: " +
-                              std::to_string(index.file_number));
-  return file_cache_->Get(options, sfile->file_number(), sfile->file_size(),
-                          index.blob_handle, record, value);
+  std::string cache_key;
+
+  if (blob_cache_) {
+    cache_key = EncodeBlobCache(index);
+    bool cache_hit;
+    Status s = TryGetBlobCache(cache_key, record, value, &cache_hit);
+    if (!s.ok()) return s;
+    if (cache_hit) return s;
+  }
+
+  OwnedSlice blob;
+  Status s = file_cache_->Get(options, index.file_number, index.blob_handle,
+                              record, &blob);
+  if (!s.ok()) {
+    return s;
+  }
+
+  if (blob_cache_ && options.fill_cache) {
+    Cache::Handle* cache_handle = nullptr;
+    auto cache_value = new OwnedSlice(std::move(blob));
```

Comment:
Maybe using `blob.release()`?

Reply:
it returns `char *` which is not what OwnedSlicce wants

Reply:
I mean we can pass char* to `blob_cache_->Insert()`?

Reply:
We can't because we need a deleter `char*` which is implemented by `OwnedSlice`. The other place does the same way